src/main/java/com/aws/greengrass/componentmanager/ComponentManager.java

```diff
@@ -512,6 +546,124 @@ public void cleanupStaleVersions() {
         logger.atDebug("cleanup-stale-versions-finish").log();
     }
 
+    /**
+     * Uninstall stale APK packages from Android core.
+     */
+    public void uninstallStaleAndroidPackages() {
+        logger.atDebug("uninstall-stale-android-packages-start").log();
+        Set<String> androidPackagesToKeep = getAndroidPackagesToKeep();
+        Set<String> packagesToRemove = listAndroidPackagesToRemove(androidPackagesToKeep);
+
+        AndroidPackageManager androidPackageManager = platform.getAndroidPackageManager();
+        for (String packageToRemove : packagesToRemove) {
+            if (androidPackageManager == null) {
+                // even if no APK manager we will mark package as uninstalled for testing purposes
+                updateAPKInstalled(packageToRemove, false);
+            } else {
+                Future future = executorService.submit(() -> {
+                    try {
+                        androidPackageManager.uninstallPackage(packageToRemove, null);
+                        updateAPKInstalled(packageToRemove, false);
+                    } catch (IOException | InterruptedException e) {
+                        logger.atError().kv(COMPONENT_NAME, packageToRemove).setCause(e)
+                                .log("Failed to uninstall Android package");
+                    }
+                    });
+                try {
+                    future.get(DEFAULT_ANDROID_PACKAGE_UNINSTALL_MS, TimeUnit.MILLISECONDS);
+                } catch (TimeoutException | ExecutionException | InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)